service-base/src/main/java/org/eclipse/hono/service/impl/RequestResponseEndpoint.java

```diff
@@ -0,0 +1,147 @@
+ */
+package org.eclipse.hono.service.impl;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.eventbus.MessageConsumer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonSender;
+import org.apache.qpid.proton.amqp.transport.AmqpError;
+import org.apache.qpid.proton.codec.DecodeException;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.config.ServiceConfigProperties;
+import org.eclipse.hono.service.amqp.BaseEndpoint;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.ResourceIdentifier;
+
+import java.util.Objects;
+
+import static io.vertx.proton.ProtonHelper.condition;
+
+/**
+ * An abstract base class for implementing endpoints that implement a request response pattern.
+ * <p>
+ * It is used e.g. in the implementation of the device registration and the credentials API endpoints.
+ */
+public abstract class RequestResponseEndpoint<T extends ServiceConfigProperties> extends BaseEndpoint<T> {
+    public static final int REQUEST_RESPONSE_ENDPOINT_DEFAULT_CREDITS = 20;
+
+    private int prefetchReceiver = REQUEST_RESPONSE_ENDPOINT_DEFAULT_CREDITS;
+
+    /**
+     * Creates an endpoint for a Vertx instance.
+     *
+     * @param vertx The Vertx instance to use.
+     * @throws NullPointerException if vertx is {@code null};
+     */
+    protected RequestResponseEndpoint(final Vertx vertx) {
+        super(Objects.requireNonNull(vertx));
+    }
+
+    /**
+     * Checks whether a given message contains all required properties.
+     *
+     * @param linkTarget The resource path to check the message's properties against for consistency.
+     * @param msg The AMQP 1.0 message to perform the checks on.
+     * @return {@code true} if the message passes all checks.
+     */
+    protected abstract boolean verifyMessage(final ResourceIdentifier linkTarget, final Message msg);
+
+    /**
+     * Process the received AMQP message.
+     *
+     * @param message The Message to process.
+     */
+    protected abstract void processRequest(final Message message);
+
+    /**
+     * Construct an AMQP reply message that is send back to the caller.
+     *
+     * @param message The Message as JsonObject from which a reply message is constructed.
+     */
+    protected abstract Message getAmqpReply(final io.vertx.core.eventbus.Message<JsonObject> message);
+
+    /**
+     * Gets the credits this endpoint grants.
+     *
+     * @return The amount of credits granted.
+     */
+    public final int getPrefetchReceiver() {
+        return prefetchReceiver;
+    }
+
+    /**
+     * Set the credits this endpoint shall grant.
+     */
+    protected final void setPrefetchReceiver(final int prefetchReceiver) {
+        this.prefetchReceiver = prefetchReceiver;
+    }
+
+    @Override
+    public void onLinkAttach(final ProtonReceiver receiver, final ResourceIdentifier targetAddress) {
```

Comment:
is this method supposed to be overridden? if so, please document what this method currently does and suggest to subclasses to override it. Otherwise, this method should be final.